components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/ico/IconDirectoryEntry.kt

```diff
@@ -0,0 +1,335 @@
+package mozilla.components.browser.icons.decoder.ico
+
+import android.graphics.Bitmap
+import mozilla.components.browser.icons.decoder.HEADER_LENGTH_BYTES
+import mozilla.components.browser.icons.decoder.ICON_DIRECTORY_ENTRY_LENGTH_BYTES
+import mozilla.components.browser.icons.decoder.IconDecoder
+import mozilla.components.browser.icons.decoder.ZERO_BYTE
+import mozilla.components.support.ktx.kotlin.containsAtOffset
+import mozilla.components.support.ktx.kotlin.toBitmap
+
+const val MAX_BITS_PER_PIXEL = 32
+
+@Suppress("MagicNumber")
+internal data class IconDirectoryEntry(
+    val width: Int,
+    val height: Int,
+    val paletteSize: Int,
+    val bitsPerPixel: Int,
+    val payloadSize: Int,
+    val payloadOffset: Int,
+    val payloadIsPNG: Boolean,
+    val directoryIndex: Int
+) : Comparable<IconDirectoryEntry> {
+    @Suppress("ReturnCount", "ComplexMethod")
+    override fun compareTo(other: IconDirectoryEntry): Int {
+        if (width > other.width) {
+            return 1
+        }
+
+        if (width < other.width) {
+            return -1
+        }
+
+        // Where both images exceed the max BPP, take the smaller of the two BPP values.
+        if (bitsPerPixel >= MAX_BITS_PER_PIXEL && other.bitsPerPixel >= MAX_BITS_PER_PIXEL) {
+            if (bitsPerPixel < other.bitsPerPixel) {
+                return 1
+            }
+
+            if (bitsPerPixel > other.bitsPerPixel) {
+                return -1
+            }
+        }
+
+        // Otherwise, take the larger of the BPP values.
+        if (bitsPerPixel > other.bitsPerPixel) {
```

Comment:
Can the `if` statements be converted to a `when` statement?

Reply:
That's a great idea. This code was a 1:1 conversion from Fennec and a `when` statement should be much nicer to read! I'll update the PR.

Reply:
👍